File: mcp-server/internal/domains/agents/apply.go
package agents

import (
	"context"
	"fmt"

	"github.com/stigmer/stigmer/mcp-server/internal/domains"
	agentv1 "github.com/stigmer/stigmer/mcp-server/proto/ai/stigmer/agentic/agent/v1"
	"google.golang.org/grpc"
)

// Apply creates or updates an agent via the AgentCommandController.Apply RPC.
// The agent proto must be fully constructed (use AgentInput.ToProto()).
func Apply(ctx context.Context, serverAddress string, agent *agentv1.Agent) (string, error) {
	return domains.WithConnection(ctx, serverAddress,
		func(ctx context.Context, conn *grpc.ClientConn) (string, error) {
			client := agentv1.NewAgentCommandControllerClient(conn)
			result, err := client.Apply(ctx, agent)
			if err != nil {
				desc := fmt.Sprintf("agent %q in org %q", agent.GetMetadata().GetSlug(), agent.GetMetadata().GetOrg())
				return "", domains.RPCError(err, desc)
			}
			return domains.MarshalJSON(result)
		})
}

File: mcp-server/src/domains/agents/apply.ts
// Agent apply path: create-or-update via the AgentCommandController.apply RPC.
// The flat MCP input is projected into a fully-formed Agent proto by the
// generated agentInputToProto bridge (codegen, src/gen/agent.ts) before the call.
// Go parity: mcp-server/internal/domains/agents/apply.go.

import { AgentSchema } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/api_pb";
import { AgentCommandController } from "@stigmer/protos/ai/stigmer/agentic/agent/v1/command_pb";

import { agentInputToProto, type AgentInput } from "../../gen/agent.js";
import { withClient } from "../client.js";
import { toProtoJson } from "../marshal.js";
import { rpcError } from "../rpcerr.js";

/** Create or update an agent, returning the persisted agent as protojson. */
export async function applyAgent(
  serverAddress: string,
  token: string,
  input: AgentInput,
): Promise<string> {
  const agent = agentInputToProto(input);
  const desc = `agent "${agent.metadata?.slug ?? ""}" in org "${agent.metadata?.org ?? ""}"`;
  return withClient(AgentCommandController, serverAddress, token, async (client, callOptions) => {
    try {
      const result = await client.apply(agent, callOptions);
      return toProtoJson(AgentSchema, result);
    } catch (err) {
      throw rpcError(err, desc);
    }
  });
}
